src/Compilers/CSharp/Portable/Compilation/MemberSemanticModel.cs

```diff
@@ -773,7 +773,9 @@ public override ForEachStatementInfo GetForEachStatementInfo(ForEachStatementSyn
 
         public override ForEachStatementInfo GetForEachStatementInfo(CommonForEachStatementSyntax node)
         {
-            BoundForEachStatement boundForEach = (BoundForEachStatement)GetUpperBoundNode(node);
+            // Declaration expressions can result in having a BoundBlock associated with a given foreach statement,
+            // in addition to a BoundForEachStatement
+            BoundForEachStatement boundForEach = GetBoundNodes(node).OfType<BoundForEachStatement>().FirstOrDefault();
```

Comment:
I am not sure whether this is the right place for the fix. The block added for expression locals is supposed to be marked as compiler generated, it feels unexpected that GetUpperBoundNode/GetBoundNodes returns it. It looks like this bug gives us definitive answer for #10604 and we should change code in ShouldAddNode instead.